source/kernel/src/registry/mod.rs

```diff
@@ -0,0 +1,440 @@
+    // and sent over the bbq::MpscProducer
+    Userspace {
+        nonce: u32,
+        outgoing: bbq::MpscProducer,
+    },
+}
+
+/// A UserspaceHandle is used to process incoming serialized messages from
+/// userspace. It contains a method that can be used to deserialize messages
+/// from a given UUID, and send that request (if the deserialization is
+/// successful) to a given driver service.
+pub struct UserspaceHandle {
+    req_producer_leaked: LeakedKProducer,
+    req_deser: ErasedDeserHandler,
+}
+
+/// A KernelHandle is used to send typed messages to a kernelspace Driver
+/// service.
+pub struct KernelHandle<RD: RegisteredDriver> {
+    prod: KProducer<Message<RD>>,
+}
+
+type ErasedDeserHandler =
+    unsafe fn(UserRequest<'_>, &LeakedKProducer, &bbq::MpscProducer) -> Result<(), ()>;
+
+/// The payload of a registry item.
+///
+/// The typeid is stored here to allow the userspace handle to look up the UUID key
+/// without knowing the proper typeid. Kernel space drivers should always check that the
+/// tuple type id is correct.
+struct RegistryValue {
+    req_resp_tuple_id: TypeId,
+    req_prod_leaked: LeakedKProducer,
+    req_deser: Option<ErasedDeserHandler>,
+}
+
+/// Right now we don't use a real HashMap, but rather a hand-rolled index map.
+/// Therefore our registry is basically a `Vec<RegistryItem>`.
+struct RegistryItem {
+    key: Uuid,
+    value: RegistryValue,
+}
+
+// Registry
+
+impl Registry {
+    /// Create a new registry with room for up to `max_items` registered drivers.
+    pub fn new(guard: &mut HeapGuard, max_items: usize) -> Self {
+        Self {
+            items: guard.alloc_fixed_vec(max_items).map_err(drop).unwrap(),
+        }
+    }
+
+    /// Register a driver service ONLY for use in the kernel, including drivers.
+    ///
+    /// Driver services registered with [Registry::set_konly] can NOT be queried
+    /// or interfaced with from Userspace. If a registered service has request
+    /// and response types that are serializable, it can instead be registered
+    /// with [Registry::set] which allows for userspace access.
+    pub fn set_konly<RD: RegisteredDriver>(
+        &mut self,
+        kch: &KProducer<Message<RD>>,
+    ) -> Result<(), ()> {
+        if self.items.iter().any(|i| i.key == RD::UUID) {
+            return Err(());
+        }
+        self.items
+            .push(RegistryItem {
+                key: RD::UUID,
+                value: RegistryValue {
+                    req_resp_tuple_id: RD::type_id(),
+                    req_prod_leaked: kch.clone().leak_erased(),
+                    req_deser: None,
+                },
+            })
+            .map_err(drop)
+    }
+
+    /// Register a driver service for use in the kernel (including drivers) as
+    /// well as in userspace.
+    ///
+    /// See [Registry::set_konly] if the request and response types are not
+    /// serializable.
+    pub fn set<RD>(&mut self, kch: &KProducer<Message<RD>>) -> Result<(), ()>
```

Comment:
naming nit, take it or leave it: the name `set` (and `set_konly`) kind of imply to me that it can overwrite a previous value, but this fails if a driver is already registered. maybe consider `register` and `register_konly`, instead, so that it doesn't sound like a map where a value can be set multiple times?